src/extension/serializer.ts

```diff
@@ -186,7 +187,17 @@ export abstract class SerializerBase implements NotebookSerializer, Disposable {
     const cells = await SerializerBase.addExecInfo(data, this.kernel)
 
     const metadata = data.metadata
-    data = new NotebookData(cells)
+
+    // Prune any ghost cells when saving.
+    let cellsToSave = []
```

Comment:
nit, this should be const. what's constant is the handle (unassigable) to the array whereas it's elements are mutable.

Reply:
drastic modification of the markdown document at serialization time will likely cause a reload in vscode which wipes editor state. wondering if `onWillSave...` allows to use the WorkspaceEdit APIs to make this transparent to the user/UX.

Reply:
this looks promising; we might have to bump to a higher minimum VS Code version (which is fine) to get this API:

https://github.com/microsoft/vscode/blob/1.92.1/src/vscode-dts/vscode.d.ts#L13519

Reply:
> drastic modification of the markdown document at serialization time will likely cause a reload in vscode 

I'm not sure I fully follow you about the problem and the proposed solution.

My thinking was that we want the in memory/rendered version to be different from the on disk version. So ghost cells should appear in the memory version but not the on disk representation.

I thought the way vscode serialization worked was VSCode invoked the serializer with the data NoteBookCellData  structure representing the data and then the serializer gets back a stream of bytes which vscode then writes to the file.  So I was thinking that as part of the serializer we could filter out the ghost cells.  It sounds like your saying that VSCode can detect if the serializer modifies the byte stream and if it does trigger a reload?

Maybe I just implemented it at the wrong place? For example, if we implemented the filtering in the golang serializer how would vscode know that the byte stream it was getting back.

How would using onWillSave and WorkspaceEdit solve this problem? If we use onWillSave and WorkspaceEdit to remove the ghost cells I think the removal of the deletion of the ghost cells would be visible to the users? What happens in the "auto save" case. Consider the following use case.

1) I start editing a markdown cell
2) Ghost cells are rendered
3) I save the file  or file gets auto saved

I think the desired behavior is
i) serialized notebook doesn't contain the ghost cells
ii) ghost cells continue to show up in the UI

Reply:
gotcha. i think i may have misunderstood and taken the "save" to literal. In my mind "save" is a user action whereas serialization is scheduled by vscode and not exclusively triggered by saving a dirty document. thanks for the clarification, i know understand the purpose of the code is to skip serializing ghost cells unless they were previously accepted.